drivers/network/netio/netio.c

```diff
@@ -0,0 +1,860 @@
+WskBind(_In_ PWSK_SOCKET Socket, _In_ PSOCKADDR LocalAddress, _Reserved_ ULONG Flags, _Inout_ PIRP Irp)
+{
+    NTSTATUS status;
+    struct _WSK_SOCKET_INTERNAL *s = (struct _WSK_SOCKET_INTERNAL *)Socket;
+    PTRANSPORT_ADDRESS ta = TdiTransportAddressFromSocketAddress(LocalAddress);
+
+    IoSetNextIrpStackLocation(Irp);
+
+    if (ta == NULL)
+    {
+        Irp->IoStatus.Status = STATUS_INSUFFICIENT_RESOURCES;
+        status = STATUS_INSUFFICIENT_RESOURCES;
+        goto err_out;
+    }
+    if (s->LocalAddressHandle != NULL)
+    {
+        ZwClose(s->LocalAddressHandle);
+        s->LocalAddressHandle = NULL;
+        s->LocalAddressFile = NULL;
+    }
+
+    status = TdiOpenAddressFile(&s->TdiName,
+                                ta, AFD_SHARE_REUSE, &s->LocalAddressHandle, &s->LocalAddressFile);
+    if (NT_SUCCESS(status))
+    {
+        memcpy(&s->LocalAddress, LocalAddress, sizeof(s->LocalAddress));
+    }
+    ExFreePoolWithTag(ta, 'ADDR');
+
+err_out:
+    Irp->IoStatus.Status = status;
+    IoCompleteRequest(Irp, IO_NETWORK_INCREMENT);
+
+    return STATUS_PENDING;
+}
+
+enum direction
+{
+    DIR_SEND,
+    DIR_RECEIVE
+};
+
+static NTSTATUS WSKAPI
+WskSendTo(
+    _In_ PWSK_SOCKET Socket,
+    _In_ PWSK_BUF Buffer,
+    _Reserved_ ULONG Flags,
+    _In_opt_ PSOCKADDR RemoteAddress,
+    _In_ ULONG ControlInfoLength,
+    _In_reads_bytes_opt_(ControlInfoLength) PCMSGHDR ControlInfo,
+    _Inout_ PIRP Irp)
+{
+    PIRP tdiIrp = NULL;
+    struct _WSK_SOCKET_INTERNAL *s = (struct _WSK_SOCKET_INTERNAL *)Socket;
+    PTDI_CONNECTION_INFORMATION TargetConnectionInfo;
+    NTSTATUS status;
+    void *BufferData;
+    struct NetioContext *nc;
+
+    IoSetNextIrpStackLocation(Irp);
+
+    status = STATUS_INSUFFICIENT_RESOURCES;
+    nc = ExAllocatePoolWithTag(NonPagedPool, sizeof(*nc), 'NEIO');
+    if (nc == NULL)
+    {
+        goto err_out;
+    }
+    nc->socket = s;
+    nc->UserIrp = Irp;
+
+    TargetConnectionInfo = TdiConnectionInfoFromSocketAddress(RemoteAddress);
+    if (TargetConnectionInfo == NULL)
+    {
+        goto err_out_free_nc;
+    }
+
+	/* TODO: @thomasfabber: how do I undo this in case the
+         * TdiSendDatagram routine fails? Or does it handle
+         * that gracefully?
+         */
+
+    BufferData = MmGetSystemAddressForMdlSafe(Buffer->Mdl, NormalPagePriority);
+    if (BufferData == NULL)
+    {
+        DbgPrint("Error mapping MDL\n");
+        goto err_out_free_nc_and_tci;
+    }
+
+    IoMarkIrpPending(Irp);
+    SocketGet(s);
+
+    /* This will create a tdiIrp: */
+    status = TdiSendDatagram(&tdiIrp, s->LocalAddressFile, ((char *)BufferData) + Buffer->Offset,
+                        Buffer->Length, TargetConnectionInfo, NetioComplete, nc);
+
+    /* If allocating tdiIrp fails we get here.
+     * Call the IoCompletion of the application's Irp so this Irp
+     * gets freed.
+     */
+    if (!NT_SUCCESS(status))
+    {
+        SocketPut(s);
+        goto err_out_free_nc_and_tci_unmap;
```

Comment:
This error handling looks good now. But one more leak: you're still responsible for freeing `TargetConnectionInfo` in the success case. So it should probably become a member of `nc`.

Reply:
OK, done.

My apologies for the leak ...